internal/querynode/query_collection.go

```diff
@@ -141,34 +142,54 @@ func (q *queryCollection) close() {
 	q.globalSegmentManager.close()
 }
 
+func (q *queryCollection) getCollectionInReplica(collectionID int64) (*Collection, error) {
+	streamingCollection, err := q.streaming.replica.getCollectionByID(collectionID)
+	if err != nil {
+		log.Warn("Cannot find collection in streaming replica:", zap.Int64("collectionID", collectionID))
+	} else {
+		return streamingCollection, nil
+	}
+	historicalCollection, err := q.historical.replica.getCollectionByID(collectionID)
+	if err != nil {
+		log.Warn("Cannot find collection in historical replica:", zap.Int64("collectionID", collectionID))
+	} else {
+		return historicalCollection, nil
+	}
+	return nil, fmt.Errorf("Cannot find collection in both streaming and historical replica. collectionID = %d", collectionID)
+}
+
 // registerCollectionTSafe registers tSafe watcher if vChannels exists
 func (q *queryCollection) registerCollectionTSafe() error {
-	collection, err := q.streaming.replica.getCollectionByID(q.collectionID)
-	if err != nil {
-		return err
+	streamingCollection, err1 := q.streaming.replica.getCollectionByID(q.collectionID)
+
+	if err1 == nil {
+		for _, channel := range streamingCollection.getVChannels() {
+			err := q.addTSafeWatcher(channel)
+			if err != nil {
+				return err
+			}
+		}
+		log.Debug("register tSafe watcher and init watcher select case",
+			zap.Any("collectionID", streamingCollection.ID()),
+			zap.Any("dml channels", streamingCollection.getVChannels()))
 	}
-	// historicalCollection, err := q.historical.replica.getCollectionByID(q.collectionID)
-	// if err != nil {
-	// 	return err
-	// }
 
-	log.Debug("register tSafe watcher and init watcher select case",
-		zap.Any("collectionID", collection.ID()),
-		zap.Any("dml channels", collection.getVChannels()),
-		// zap.Any("delta channels", collection.getVChannels()),
-	)
-	for _, channel := range collection.getVChannels() {
-		err = q.addTSafeWatcher(channel)
-		if err != nil {
-			return err
+	historicalCollection, err2 := q.historical.replica.getCollectionByID(q.collectionID)
+	if err2 == nil {
+		for _, channel := range historicalCollection.getVChannels() {
+			err := q.addTSafeWatcher(channel)
+			if err != nil {
+				return err
+			}
 		}
+		log.Debug("register tSafe watcher and init watcher select case",
+			zap.Any("collectionID", historicalCollection.ID()),
+			zap.Any("delta channels", historicalCollection.getVChannels()))
+	}
+	if err1 != nil && err2 != nil {
+		return fmt.Errorf("collection cannot be found in both historical and streaming replica")
```

Comment:
I think we should check if collection is exists in both streaming and historical.
Collection would be created in both streaming and historical even though there is dm data in streaming and no dm data in historical.

Reply:
> I think we should check if collection is exists in both streaming and historical. Collection would be created in both streaming and historical even though there is dm data in streaming and no dm data in historical.

It is not necessary to check whether there is this collection in historical or streaming. You can call queryCollection.addTSafeWatcher to add vchannel when you watchDmchane/deltaChannel after watchQueryChannel.